File: mobile/lib/src/services/user/message.dart
import 'dart:convert';

import 'package:http/http.dart';
import 'package:revelationsai/src/constants/api.dart';
import 'package:revelationsai/src/models/pagination.dart';
import 'package:revelationsai/src/models/search.dart';
import 'package:revelationsai/src/models/user/message.dart';
import 'package:revelationsai/src/utils/http_helpers.dart';

class UserMessageService {
  static Future<PaginatedEntitiesResponseData<UserMessage>> getUserMessages({
    PaginatedEntitiesRequestOptions paginationOptions =
        const PaginatedEntitiesRequestOptions(),
    required String session,
  }) async {
    Response res = await get(
      Uri.parse('${API.url}/user-messages?${paginationOptions.searchQuery}'),
      headers: <String, String>{
        'Authorization': 'Bearer $session',
        'Content-Type': 'application/json; charset=UTF-8',
      },
    );

    if (!res.ok) {
      throw res.exception;
    }

    final data = jsonDecode(utf8.decode(res.bodyBytes));
    return PaginatedEntitiesResponseData.fromJson(data, (json) {
      return UserMessage.fromJson(json as Map<String, dynamic>);
    });
  }

  static Future<UserMessage> getUserMessage({
    required String id,
    required String session,
  }) async {
    Response res = await get(
      Uri.parse('${API.url}/user-messages/$id'),
      headers: <String, String>{
        'Authorization': 'Bearer $session',
        'Content-Type': 'application/json; charset=UTF-8',
      },
    );

    if (!res.ok) {
      throw res.exception;
    }

    final data = jsonDecode(utf8.decode(res.bodyBytes));
    return UserMessage.fromJson(data);
  }

  static Future<PaginatedEntitiesResponseData<UserMessage>>
      searchForUserMessages({
    PaginatedEntitiesRequestOptions paginationOptions =
        const PaginatedEntitiesRequestOptions(),
    required Query query,
    required String session,
  }) async {
    Response res = await post(
      Uri.parse(
        '${API.url}/user-messages/search?${paginationOptions.searchQuery}',
      ),
      headers: <String, String>{
        'Authorization': 'Bearer $session',
        'Content-Type': 'application/json; charset=UTF-8',
      },
      body: jsonEncode(query.toJson()),
    );

    if (!res.ok) {
      throw res.exception;
    }

    final data = jsonDecode(utf8.decode(res.bodyBytes));
    return PaginatedEntitiesResponseData.fromJson(data, (json) {
      return UserMessage.fromJson(json as Map<String, dynamic>);
    });
  }
}

File: packages/client/src/services/user/message.ts
import type { UserMessage } from '@revelationsai/core/model/user/message';
import apiConfig from '../../configs/api';
import { GetEntitiesSearchParams } from '../helpers/search-params';
import type {
  PaginatedEntitiesOptions,
  PaginatedEntitiesResponse,
  ProtectedApiOptions,
  SearchForEntitiesOptions
} from '../types';

export async function getUserMessages(options: PaginatedEntitiesOptions & ProtectedApiOptions) {
  const searchParams = GetEntitiesSearchParams(options);
  const response = await fetch(`${apiConfig.url}/user-messages?${searchParams.toString()}`, {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${options.session}`
    }
  });

  if (!response.ok) {
    console.error(`Error retrieving user messages. Received response:`, JSON.stringify(response));
    const data = (await response.json()) as { error?: string };
    throw new Error(data.error || 'Error retrieving user messages.');
  }

  const { entities, page, perPage } =
    (await response.json()) as PaginatedEntitiesResponse<UserMessage>;

  return {
    userMessages: entities,
    page,
    perPage
  };
}

export async function getUserMessage(id: string, options: ProtectedApiOptions) {
  const response = await fetch(`${apiConfig.url}/user-messages/${id}`, {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${options.session}`
    }
  });

  if (!response.ok) {
    console.error(`Error retrieving user message. Received response:`, JSON.stringify(response));
    const data = (await response.json()) as { error?: string };
    throw new Error(data.error || 'Error retrieving user message.');
  }

  const userMessage = (await response.json()) as UserMessage;

  return userMessage;
}

export async function searchForUserMessages(
  options: SearchForEntitiesOptions & PaginatedEntitiesOptions & ProtectedApiOptions
) {
  const searchParams = GetEntitiesSearchParams(options);
  const response = await fetch(`${apiConfig.url}/user-messages/search?${searchParams.toString()}`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${options.session}`
    },
    body: JSON.stringify(options.query)
  });

  if (!response.ok) {
    console.error(
      `Error searching for user messages. Received response:`,
      JSON.stringify(response)
    );
    const data = (await response.json()) as { error?: string };
    throw new Error(data.error || 'Error searching for user messages.');
  }

  const { entities, page, perPage } =
    (await response.json()) as PaginatedEntitiesResponse<UserMessage>;

  return {
    userMessages: entities,
    page,
    perPage
  };
}
